src/predictor.rs

```diff
@@ -0,0 +1,753 @@
+            PlanarConfiguration::Chunky => {
+                self.bits_per_sample as usize * self.samples_per_pixel as usize
+            }
+            PlanarConfiguration::Planar => self.bits_per_sample as usize,
+        }
+    }
+
+    /// The number of chunks in the horizontal (x) direction
+    fn chunks_across(&self) -> u32 {
+        self.image_width.div_ceil(self.chunk_width)
+    }
+
+    /// The number of chunks in the vertical (y) direction
+    fn chunks_down(&self) -> u32 {
+        self.image_height.div_ceil(self.chunk_height)
+    }
+}
+
+/// reverse horizontal predictor
+///
+/// fixes byte order before reversing differencing
+pub(crate) fn unpredict_hdiff(
+    buffer: Bytes,
+    predictor_info: &PredictorInfo,
+    tile_x: u32,
+) -> AsyncTiffResult<Bytes> {
+    let output_row_stride = predictor_info.output_row_stride(tile_x)?;
+    let samples = predictor_info.samples_per_pixel as usize;
+    let bit_depth = predictor_info.bits_per_sample;
+
+    let buffer = fix_endianness(buffer, predictor_info.endianness, bit_depth);
+    let mut res = BytesMut::from(buffer);
+    for buf in res.chunks_mut(output_row_stride) {
+        rev_hpredict_nsamp(buf, bit_depth, samples);
+    }
+    Ok(res.into())
+}
+
+/// Reverse predictor convenience function for horizontal differencing
+///
+/// From image-tiff
+///
+/// This should be used _after_ endianness fixing
+pub fn rev_hpredict_nsamp(buf: &mut [u8], bit_depth: u16, samples: usize) {
+    match bit_depth {
+        0..=8 => {
+            for i in samples..buf.len() {
+                buf[i] = buf[i].wrapping_add(buf[i - samples]);
+            }
+        }
+        9..=16 => {
+            for i in (samples * 2..buf.len()).step_by(2) {
+                let v = u16::from_ne_bytes(buf[i..][..2].try_into().unwrap());
+                let p = u16::from_ne_bytes(buf[i - 2 * samples..][..2].try_into().unwrap());
+                buf[i..][..2].copy_from_slice(&(v.wrapping_add(p)).to_ne_bytes());
+            }
+        }
+        17..=32 => {
+            for i in (samples * 4..buf.len()).step_by(4) {
+                let v = u32::from_ne_bytes(buf[i..][..4].try_into().unwrap());
+                let p = u32::from_ne_bytes(buf[i - 4 * samples..][..4].try_into().unwrap());
+                buf[i..][..4].copy_from_slice(&(v.wrapping_add(p)).to_ne_bytes());
+            }
+        }
+        33..=64 => {
+            for i in (samples * 8..buf.len()).step_by(8) {
+                let v = u64::from_ne_bytes(buf[i..][..8].try_into().unwrap());
+                let p = u64::from_ne_bytes(buf[i - 8 * samples..][..8].try_into().unwrap());
+                buf[i..][..8].copy_from_slice(&(v.wrapping_add(p)).to_ne_bytes());
+            }
+        }
+        _ => {
+            unreachable!("Caller should have validated arguments. Please file a bug.")
+        }
+    }
+}
+
+/// Fix endianness. If `byte_order` matches the host, then conversion is a no-op.
+///
+/// from image-tiff
```

Comment:
Do you want this "from image-tiff" line to show up in the docs? Or only visible as a code comment?

```suggestion
// from image-tiff
```